src/main/java/com/cloudbees/hudson/plugins/folder/AbstractFolder.java

```diff
@@ -0,0 +1,746 @@
+        checkPermission(CONFIGURE);
+
+        req.setCharacterEncoding("UTF-8");
+        JSONObject json = req.getSubmittedForm();
+
+        BulkChange bc = new BulkChange(this);
+        try {
+            description = json.getString("description");
+            displayName = Util.fixEmpty(json.optString("displayNameOrNull"));
+            if (json.has("viewsTabBar")) {
+                viewsTabBar = req.bindJSON(ViewsTabBar.class, json.getJSONObject("viewsTabBar"));
+            }
+
+            if (json.has("primaryView")) {
+                primaryView = json.getString("primaryView");
+            }
+
+            properties.rebuild(req, json, getDescriptor().getPropertyDescriptors());
+            for (AbstractFolderProperty p : properties) {
+                p.setOwner(this);
+            }
+
+            healthMetrics.replaceBy(req.bindJSONToList(FolderHealthMetric.class, json.get("healthMetrics")));
+
+            icon = req.bindJSON(FolderIcon.class, req.getSubmittedForm().getJSONObject("icon"));
+            icon.setOwner(this);
+
+            submit(req, rsp);
+
+            save();
+            bc.commit();
+        } finally {
+            bc.abort();
+        }
+
+        // TODO boilerplate
+        String newName = json.getString("name");
+        ProjectNamingStrategy namingStrategy = Jenkins.getActiveInstance().getProjectNamingStrategy();
+        if (newName != null && !newName.equals(name)) {
+            // check this error early to avoid HTTP response splitting.
+            Jenkins.checkGoodName(newName);
+            namingStrategy.checkName(newName);
+            rsp.sendRedirect("rename?newName=" + URLEncoder.encode(newName, "UTF-8"));
+        } else {
+            if (namingStrategy.isForceExistingJobs()) {
+                namingStrategy.checkName(name);
+            }
+            FormApply.success(".").generateResponse(req, rsp, this);
+        }
+    }
+
+    protected void submit(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException, Descriptor.FormException {}
+
+    // TODO boilerplate like this should not be necessary: JENKINS-22936
+    @RequirePOST
+    public void doDoRename(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException {
+
+        if (!hasPermission(CONFIGURE)) {
+            // rename is essentially delete followed by a create
+            checkPermission(CREATE);
+            checkPermission(DELETE);
+        }
+
+        String newName = req.getParameter("newName");
+        Jenkins.checkGoodName(newName);
+
+        String blocker = renameBlocker();
+        if (blocker != null) {
+            rsp.sendRedirect("rename?newName=" + URLEncoder.encode(newName, "UTF-8") + "&blocker=" + URLEncoder.encode(blocker, "UT-8"));
+            return;
+        }
+
+        renameTo(newName);
+        rsp.sendRedirect2("../" + newName);
```

Comment:
because this is from `.../jobs/oldName/rename` and we want to redirect to `.../jobs/newName` not `.../jobs/oldName/newName`